packages/plugins/typescript/graphql-apollo/src/visitor.ts

```diff
@@ -0,0 +1,107 @@
+import {
+  ClientSideBasePluginConfig,
+  ClientSideBaseVisitor,
+  LoadedFragment,
+} from '@graphql-codegen/visitor-plugin-common';
+import autoBind from 'auto-bind';
+import { camelCase } from 'change-case-all';
+import { GraphQLSchema, Kind, OperationDefinitionNode, print } from 'graphql';
+import { RawGraphQLApolloPluginConfig } from './config';
+
+export interface GraphQLRequestPluginConfig extends ClientSideBasePluginConfig {
+  rawRequest: boolean;
+}
+
+export class GraphQLApolloVisitor extends ClientSideBaseVisitor<
+  RawGraphQLApolloPluginConfig,
+  GraphQLRequestPluginConfig
+> {
+  private _operationsToInclude: {
+    node: OperationDefinitionNode;
+    documentVariableName: string;
+    operationType: string;
+    operationResultType: string;
+    operationVariablesTypes: string;
+  }[] = [];
+
+  constructor(schema: GraphQLSchema, fragments: LoadedFragment[], rawConfig: RawGraphQLApolloPluginConfig) {
+    super(schema, fragments, rawConfig, {});
+
+    autoBind(this);
+
+    const typeImport = this.config.useTypeImports ? 'import type' : 'import';
+
+    this._additionalImports.push(`${typeImport} { ApolloClient } from '@apollo/client';`);
+  }
+
+  public OperationDefinition(node: OperationDefinitionNode) {
+    const operationName = node.name?.value;
+
+    if (!operationName) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        `Anonymous GraphQL operation was ignored in "typescript-graphql-request", please make sure to name your operation: `,
+        print(node)
+      );
+
+      return null;
+    }
+
+    return super.OperationDefinition(node);
+  }
+
+  protected buildOperation(
+    node: OperationDefinitionNode,
+    documentVariableName: string,
+    operationType: string,
+    operationResultType: string,
+    operationVariablesTypes: string
+  ): string {
+    this._operationsToInclude.push({
+      node,
+      documentVariableName,
+      operationType,
+      operationResultType,
+      operationVariablesTypes,
+    });
+
+    return null;
+  }
+
+  public get sdkContent(): string {
+    const sdkOperations = this._operationsToInclude.map(x => {
+      const optionalVariables =
+        !x.node.variableDefinitions ||
+        x.node.variableDefinitions.length === 0 ||
+        x.node.variableDefinitions.every(v => v.type.kind !== Kind.NON_NULL_TYPE || v.defaultValue);
+
+      const documentVariableName = x.documentVariableName;
+
+      const operationName = x.node.name.value;
+      return `${camelCase(operationName)}${x.operationType}(variables${optionalVariables ? '?' : ''}: ${
+        x.operationVariablesTypes
+      }) {
+          return client.${GraphQLApolloVisitor.getApolloOperation(x.operationType)}<${
+        x.operationResultType
+      }>({variables, query: ${documentVariableName}})
```

Comment:
Hey there!

I'm currently building out a bit more robust of an apollo client call, and I'm finding that variables is insufficient, and that I'm using [context](https://www.apollographql.com/docs/react/api/core/ApolloClient/#ApolloClient.query).

Would it be possible for the API for the method be updated from:
```ts
client.query(variables: MyQueryVariables)
```
to 
```ts
client.query({ variables }:  QueryOptions<MyQueryVariables>) // or MutationOptions / SubscriptionOptions
```

This would not only solve my current need, but keep this plugin a bit more future-resilient

Reply:
I actually love this idea :slightly_smiling_face: Do you mind sending a PR on top of this? I'm swamped currently at work, I'll try to give this a try over the weekend if I happen to get any free time, but would love if you could help me out, either way this is going to happen :slightly_smiling_face:

Reply:
I think this also helps when we write middlewares, you can access the context you pass :slightly_smiling_face: if I understand correctly, forgive me if I'm wrong, haven't done client side apollo, I'm all backend graphql for now :slightly_smiling_face:

Reply:
however it doesn't seem like we're ever going to get a review from maintainers though, I'm starting to think to publish this from my account, but I don't think it's such a good idea though.

Reply:
> I think this also helps when we write middlewares, you can access the context you pass 🙂 if I understand correctly, forgive me if I'm wrong, haven't done client side apollo, I'm all backend graphql for now 🙂

You're absolutely correct about this. I'm using apollo in all 3, SSG/SSR/CSR and context is accessible in all of them. That's actually why I want this whole plugin to begin with, so that I have consistent and clean types/syntactic sugar. You're the real hero here.

I've got a crazy-busy week here too 😭 😎 .

Let me work up at least the base of this branch this week and if you have the bandwidth this weekend, you could maybe update the tests or something?

> however it doesn't seem like we're ever going to get a review from maintainers though, I'm starting to think to publish this from my account, but I don't think it's such a good idea though.

i'm optimistic that they'll review this soon =) It's not a complicated PR, so... fingers crossed.

And if in another 2 weeks, we still don't hear anything, I wonder what the way to self-publish this would be such that it would still work with the core codegen lib?

Reply:
I'll give it a stab over the coming weekend (just set a reminder on slack), :slightly_smiling_face: but if you happen to push something I can take care of tests and making these pass.